src/entropies_estimators/nearest_neighbors/Lord.jl

```diff
@@ -0,0 +1,167 @@
+
+    # Decrease allocations and speed up computations by pre-allocating.
+    # We're only dealing with matrices which in either axis has a maximum dimension of `k`,
+    # so this is still far within the realm of where StaticArrays shines.
+    # -------------------------------------------------------------------------------------
+    rs = @MVector zeros(D) # Scaled ellipsoid axis lengths
+    Λ = @MMatrix zeros(D, D) # Hyperellipsoid matrix
+
+    # C contains neighborhood-centroid-centered vectors, where
+    # `C[1]` := the centered query point
+    # `C[1 + j]` := the centered `j`-th neighbor of the query point.
+    C = MVector{k + 1, MVector{D}}(@SVector zeros(D) for i = 1:k+1)
+    # Centered neighbors need to be ordered row-wise in a matrix. We re-fill this matrix
+    # for every query point `xᵢ`
+    A = @MMatrix zeros(k+1, D)
+
+    # Precompute some factors
+    γ = gamma(1 + D/2)
+    f = N * π^(D/2)
+
+    h = 0.0
+    for (i, xᵢ) in enumerate(x)
+        neighborsᵢ = @views x[knn_idxs[i]]
+        # Center neighborhood points around mean of the neighborhood.
+        c = centroid(xᵢ, neighborsᵢ, C)
+        center_neighborhood!(C, c, xᵢ, neighborsᵢ) # put centered vectors in `C`
+        fill_A!(A, C, D)
+
+        # SVD. The columns of Vt are the semi-axes of the ellipsoid, while Σ gives the
+        # magnitudes of the axes.
+        U, Σ, Vt = svd(A)
+        σ₁ = Σ[1]
+        ϵᵢ = last(ds[i])
+
+        # Scale semi-axis lengths to k-th neighbor distance
+        rs .= ϵᵢ .* (Σ ./ σ₁)
+        # Create matrix ellipse representation, centered at origin (fill Λ)
+        hyperellipsoid_matrix!(Λ, Vt, rs)
+        # Take the point `xᵢ` as the origin for the neighborhood, so we can check directly
+        # whether points are inside the ellipse. This happens for a point `x`
+        # whenever `xᵀΛx <= 1`.
+        nns_centered = (pt - xᵢ for pt in neighborsᵢ)
+        kᵢ = count([transpose(p) * Λ * p <= 1.0 for p in nns_centered])
+        # In the paper the point itself is always counted inside the ellipsoid,
+        # so that there is always one point present. Here we instead set the local density
+        # to zero (by just skipping the computation) if that is the case.
+        if kᵢ > 0
+            h += log(kᵢ * γ / (f * ϵᵢ^D * prod(Σ ./ σ₁)) )
+        end
+    end
+    h = - h / N
+
+    return h / log(e.base, ℯ)
+end
+entropy(est::Lord, args...) = entropy(Shannon(), est, args...)
+
+function fill_A!(A, C, D)
+    for (j, m) in enumerate(C)
+        A[j, :] .= SVector{D}(m)
+    end
+end
+
+"""
+    center_neighborhood!(c, C, xᵢ, neighbors)
+
+Center the point `xᵢ`, as well as each of its neighboring points `nⱼ ∈ neighbors`,
+to the (precomputed) centroid `c` of the points `{xᵢ, n₁, n₂, …, nₖ}`, and store the
+centered vectors in the pre-allocated vector `C`.
+"""
+function center_neighborhood!(C::MVector{K, V}, c, xᵢ,
+        neighbors::AbstractDataset{D}) where {V, K, D}
+    rezero!(C)
+    C[1] .= xᵢ .- c
```

Comment:
This can be `C[1] = xᵢ .- c`. Assuming that both `c` and `xi` are `SVectors`, this is the fastest way to do it. If either `c` or `xi` are not SVectors, then they shouold be. I.e., `center_neighbood!` should do 0 allocations as an isolated function.